server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MDNParse.scala

```diff
@@ -0,0 +1,173 @@
+import scala.util.Try
+
+object MDNParse {
+  type UnparsedBlobIdConstraint = NonEmpty
+  type UnparsedBlobId = String Refined UnparsedBlobIdConstraint
+}
+
+object BlobIds {
+  def parse(blobIds: Seq[UnparsedBlobId]): Seq[Try[BlobId]] = {
+    blobIds.map(unparsed => BlobId.of(unparsed.value))
+  }
+}
+
+case class BlobIds(value: Seq[UnparsedBlobId])
+
+case class MDNParseRequest(accountId: AccountId,
+                           blobIds: BlobIds) extends WithAccountId {
+
+  def validate: Either[IllegalArgumentException, MDNParseRequest] = {
+    if (blobIds.value.length > 2) {
+      Left(new IllegalArgumentException(s"The number of ids requested by the client exceeds the maximum number the server is willing to process in a single method call"))
+    } else {
+      scala.Right(this)
+    }
+  }
+}
+
+object MDNNotFound {
+  def empty(): MDNNotFound = MDNNotFound(Set())
+}
+
+case class MDNNotFound(value: Set[UnparsedBlobId]) {
+  def merge(other: MDNNotFound): MDNNotFound = MDNNotFound(this.value ++ other.value)
+}
+
+object MDNNotParsable {
+  def empty(): MDNNotParsable = MDNNotParsable(Set())
+
+  def merge(p1: MDNNotParsable, p2: MDNNotParsable): MDNNotParsable = MDNNotParsable(p1.value ++ p2.value)
+}
+
+case class MDNNotParsable(value: Set[UnparsedBlobId]) {
+  def merge(other: MDNNotParsable): MDNNotParsable = MDNNotParsable(this.value ++ other.value)
+}
+
+case class MDNParseFailure(value: UnparsedBlobId)
+
+object MDNDisposition {
+  def convertFromJava(javaDisposition: JavaDisposition): MDNDisposition =
+    MDNDisposition(actionMode = javaDisposition.getActionMode.getValue,
+      sendingMode = javaDisposition.getSendingMode.getValue,
+      `type` = javaDisposition.getType.getValue)
+}
+
+case class MDNDisposition(actionMode: String,
+                          sendingMode: String,
+                          `type`: String)
+
+case class ForEmailIdField(value: String) extends AnyVal
+case class SubjectField(value: String) extends AnyVal
+case class TextBodyField(value: String) extends AnyVal
+case class ReportUAField(value: String) extends AnyVal
+case class FinalRecipientField(value: String) extends AnyVal
+case class OriginalMessageIdField(value: String) extends AnyVal
+case class ErrorField(value: String) extends AnyVal
+case class ExtensionField(value: Map[String, String]) extends AnyVal
+
+
+object MDNParsed {
+  def convertFromMDN(mdn: MDN, message: Message): MDNParsed = {
+    val report = mdn.getReport;
+    MDNParsed(
+      forEmailId = None,
+      subject = Option(SubjectField(message.getSubject)),
+      textBody = Some(TextBodyField(mdn.getHumanReadableText)),
+      reportingUA = report.getReportingUserAgentField
+        .map(userAgent => ReportUAField(s"${userAgent.getUserAgentName}; ${userAgent.getUserAgentProduct.orElse("")}"))
+        .toScala,
+      finalRecipient = FinalRecipientField(s"${report.getFinalRecipientField.getAddressType.getType}; ${report.getFinalRecipientField.getFinalRecipient.formatted()}"),
```

Comment:
Why not `finalRecipient = FinalRecipientField(eport.getFinalRecipientField.fieldValue)` so that we don't rewrite MDN logic here?

Then we modify `FinalRecipient` (java class):

```
    @Override
    public String formattedValue() {
        return FIELD_NAME + ": " + fieldValue();
    }

    private String fieldValue() {
        return addressType.getType() + "; " + finalRecipient.formatted();
    }
```